pkg/apis/metal3/v1alpha1/baremetalhost_types.go

```diff
@@ -104,6 +104,111 @@ type BMCDetails struct {
 	CredentialsName string `json:"credentialsName"`
 }
 
+// DiskType is used to specify the disk type for a logical disk, e.g. hdd or ssd.
+type DiskType string
+
+// Valid disk types for RAID configuration
+const (
+	HDD DiskType = "hdd"
+	SSD DiskType = "ssd"
+)
+
+// ValidDiskType check whether given disk type is supported or not
+func (d DiskType) ValidDiskType () bool {
+	switch d {
+	case SSD, HDD:
+		return true
+	default:
+		return false
+	}
+}
+
+// InterfaceType is used to specify the interface for a logical disk.
+type InterfaceType string
+
+// Valid interface types for RAID configuration
+const (
+	SATA InterfaceType = "sata"
+	SCSI InterfaceType = "scsi"
+	SAS  InterfaceType = "sas"
+)
+
+// ValidInterfaceType check whether given interface type is supported or not
+func (i InterfaceType) ValidInterfaceType () bool {
+	switch i {
+	case SATA, SCSI, SAS:
+		return true
+	default:
+		return false
+	}
+}
+
+// RAIDLevel type is used to specify the RAID level for a logical disk.
+type RAIDLevel string
+
+// Valid RAID levels
+const (
+	RAID0  RAIDLevel = "0"
+	RAID1  RAIDLevel = "1"
+	RAID2  RAIDLevel = "2"
+	RAID5  RAIDLevel = "5"
+	RAID6  RAIDLevel = "6"
+	RAID10 RAIDLevel = "1+0"
+	RAID50 RAIDLevel = "5+0"
+	RAID60 RAIDLevel = "6+0"
+)
+
+// ValidRAIDLevel check whether given RAID level is supported or not
+func (l RAIDLevel) ValidRAIDLevel () bool {
+	switch l {
+	case RAID0, RAID1, RAID2, RAID5, RAID6, RAID10, RAID50, RAID60:
+		return true
+	default:
+		return false
+	}
+}
+
+// RAIDVolume defines the desired configuration of volume in RAID
+type RAIDVolume struct {
+	// Size (Integer) of the logical disk to be created in GiB.  If unspecified, "MAX" will be used.
+	SizeGb *int `json:"sizeGb,omitempty"`
+
+	// RAID level for the logical disk. The following RAID levels are supported: 0, 1, 2, 5, 6, 1+0, 5+0, 6+0.
+	RAIDLevel RAIDLevel `json:"raidLevel" required:"true"`
+
+	// Name of the volume. Should be unique within the Node. If not specified, volume name will be auto-generated.
+	VolumeName string `json:"name,omitempty"`
+
+	// Set to true if this logical disk can share physical disks with other logical disks.
+	SharePhysicalDisks *bool `json:"sharePhysicalDisks,omitempty"`
+
+	// The valid values must be "hdd" or "ssd". If this is not specified, disk type will not be a criterion to find backing physical disks
+	DiskType DiskType `json:"diskType,omitempty"`
+
+	// The valid values are "sata" or "scsi" or "sas". If this is not specified, interface type will not be a criterion to find backing physical disks.
+	InterfaceType InterfaceType `json:"interfaceType,omitempty"`
+
+	// Integer, number of disks to use for the logical disk. Defaults to minimum number of disks required
+	// for the particular RAID level.
+	NumberOfPhysicalDisks int `json:"numberOfPhysicalDisks,omitempty"`
```

Comment:
When would you ever set this to anything other than the default for a given RAID mode?

Reply:
I never try to set this before.

Reply:
Well, you can have non-default number of disks, especially for RAID-0.

Reply:
OK.

Reply:
How does the validation instruction work with this field? If the user doesn't specify a value, will we get 0? But if they do specify a value then the validation ensures that it is > 0?

Reply:
Opps, I missed that point. 
In the case of user didn't specify a value, we will pass the validation of kubebuilder and got 0.
And if they specify a value, it will be validate by kubebuilder in line 129 `// +kubebuilder:validation:Minimum=1`
So I think pointer would be suitable in the case of missing this field
Thanks

Reply:
Unresolved this because it appears we decided on a pointer here but the field didn't actually change.

Also for software RAID we're supposed to be able to supply [device hints in the same format as the root device hints ](https://docs.openstack.org/ironic/latest/admin/raid.html#backing-physical-disks) we added in #495, and the DiskType and NumberOfPhysicalDisks do not apply. If we're going to keep the software RAID thing I wonder if we should separate the software and hardware RAID configurations into separate structures.

Reply:
How about this: 
```
// RAIDConfig contains the configuration that are required to config RAID in Bare Metal server
type RAIDConfig struct {
	// If true will remove existing raid configuration and set new raid configuration.
	DeleteExisting bool `json:"deleteExisting,omitempty"`

	// The list of logical disks for hardware RAID, first volume is root volume
	HVolumes []HRAIDVolume `json:"hVolumes,omitempty"`

	// The list of logical disks for software RAID, first volume is root volume, if HVolume is setted this items will be invalid.
	// The number of created Software RAID devices must be 1 or 2.
	// If there is only one Software RAID device, it has to be a RAID-1.
	// If there are two, the first one has to be a RAID-1, while the RAID level for the second one can 0, 1,
	// or 1+0. As the first RAID device will be the deployment device,
	// enforcing a RAID-1 reduces the risk of ending up with a non-booting node in case of a disk failure.
	SVolumes []SRAIDVolume
}
```